state/validation_test.go

```diff
@@ -5,74 +5,119 @@ import (
-	block := makeBlock(state, 1)
+	// A good block passes.
+	block := makeBlock(state, height)
 	err := blockExec.ValidateBlock(state, block)
 	require.NoError(t, err)
 
-	// wrong chain fails
-	block = makeBlock(state, 1)
-	block.ChainID = "not-the-real-one"
-	err = blockExec.ValidateBlock(state, block)
-	require.Error(t, err)
+	wrongHash := tmhash.Sum([]byte("this hash is wrong"))
 
-	// wrong height fails
-	block = makeBlock(state, 1)
-	block.Height += 10
-	err = blockExec.ValidateBlock(state, block)
-	require.Error(t, err)
+	// Manipulation of any header field causes failure.
+	testCases := []struct {
+		name string
+		f    func(block *types.Block)
+	}{
+		{"ChainID", func(block *types.Block) { block.ChainID = "not-the-real-one" }}, // wrong chain id
+		{"Height", func(block *types.Block) { block.Height += 10 }},                  // wrong height
+		// TODO(#2589) (#2587) : {"Time", func(block *types.Block) { block.Time.Add(-time.Second * 3600 * 24) }}, // wrong time
+		{"NumTxs", func(block *types.Block) { block.NumTxs += 10 }},     // wrong num txs
+		{"TotalTxs", func(block *types.Block) { block.TotalTxs += 10 }}, // wrong total txs
 
-	// wrong total tx fails
-	block = makeBlock(state, 1)
-	block.TotalTxs += 10
-	err = blockExec.ValidateBlock(state, block)
-	require.Error(t, err)
+		{"LastBlockID", func(block *types.Block) { block.LastBlockID.PartsHeader.Total += 10 }}, // wrong block id
+		{"LastCommitHash", func(block *types.Block) { block.LastCommitHash = wrongHash }},       // wrong commit hash
+		{"DataHash", func(block *types.Block) { block.DataHash = wrongHash }},                   // wrong data hash
 
-	// wrong blockid fails
-	block = makeBlock(state, 1)
-	block.LastBlockID.PartsHeader.Total += 10
-	err = blockExec.ValidateBlock(state, block)
-	require.Error(t, err)
+		{"ValidatorsHash", func(block *types.Block) { block.ValidatorsHash = wrongHash }},         // wrong validators hash
+		{"NextValidatorsHash", func(block *types.Block) { block.NextValidatorsHash = wrongHash }}, // wrong next validators hash
+		{"ConsensusHash", func(block *types.Block) { block.ConsensusHash = wrongHash }},           // wrong consensus hash
+		{"AppHash", func(block *types.Block) { block.AppHash = wrongHash }},                       // wrong app hash
+		{"LastResultsHash", func(block *types.Block) { block.LastResultsHash = wrongHash }},       // wrong results hash
 
-	// wrong app hash fails
-	block = makeBlock(state, 1)
-	block.AppHash = []byte("wrong app hash")
-	err = blockExec.ValidateBlock(state, block)
-	require.Error(t, err)
+		{"EvidenceHash", func(block *types.Block) { block.EvidenceHash = wrongHash }},                                    // wrong evidence hash
+		{"Proposer wrong", func(block *types.Block) { block.ProposerAddress = ed25519.GenPrivKey().PubKey().Address() }}, // wrong proposer
+		{"Proposer invalid", func(block *types.Block) { block.ProposerAddress = []byte("wrong size") }},                  // invalid proposer
```

Comment:
no need in `// invalid proposer` comment if we already have "Proposer invalid" as a test name